src/NuGet.Clients/PackageManagement.VisualStudio/ProjectSystems/ProjectNames.cs

```diff
@@ -0,0 +1,81 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using Microsoft.VisualStudio.Shell;
+
+namespace NuGet.PackageManagement.VisualStudio
+{
+    /// <summary>
+    /// Represents a project name in the solution manager.
+    /// </summary>
+    public class ProjectNames : IEquatable<ProjectNames>
+    {
+        public string FullName { get; private set; }
+        public string UniqueName { get; private set; }
+        public string ShortName { get; private set; }
+        public string CustomUniqueName { get; private set; }
+
+        public static ProjectNames FromDTEProject(EnvDTE.Project dteProject)
+        {
+            if (dteProject == null)
+            {
+                throw new ArgumentNullException(nameof(dteProject));
+            }
+
+            Debug.Assert(ThreadHelper.CheckAccess());
+
+            return new ProjectNames
+            {
+                FullName = dteProject.FullName,
+                UniqueName = EnvDTEProjectUtility.GetUniqueName(dteProject),
+                ShortName = EnvDTEProjectUtility.GetName(dteProject),
+                CustomUniqueName = EnvDTEProjectUtility.GetCustomUniqueName(dteProject)
+            };
+        }
+
+        /// <summary>
+        /// Two projects are equal if they share the same FullNames.
+        /// </summary>
+        public bool Equals(ProjectNames other)
+        {
+            if (ReferenceEquals(null, other))
```

Comment:
Why this and not 
if (other == null)
?

also elsewhere. If just a preference, fine. Discussion point.